rust/worker/src/segment/types.rs

```diff
@@ -1,29 +1,362 @@
+            user_id: None,
+            final_operation: Operation::Add,
+            metadata_to_be_merged: None,
+            final_document: None,
+            final_embedding: None,
+        }
+    }
+}
+
+impl<'a> TryFrom<(&'a OperationRecord, u32, &'a str)> for MaterializedLogRecord<'a> {
+    type Error = LogMaterializerError;
+
+    fn try_from(
+        log_operation_info: (&'a OperationRecord, u32, &'a str),
+    ) -> Result<Self, Self::Error> {
+        let log_record = log_operation_info.0;
+        let offset_id = log_operation_info.1;
+        let user_id = log_operation_info.2;
+        let metadata = match &log_record.metadata {
+            Some(metadata) => match update_metdata_to_metdata(metadata) {
+                Ok(m) => Some(m),
+                Err(e) => {
+                    return Err(LogMaterializerError::MetadataMaterializationError(e));
+                }
+            },
+            None => None,
+        };
+
+        let document = match &log_record.document {
+            Some(doc) => Some(doc.as_str()),
+            None => None,
+        };
+
+        let embedding = match &log_record.embedding {
+            Some(embedding) => Some(embedding.as_slice()),
+            None => {
+                return Err(LogMaterializerError::EmbeddingMaterializationError);
+            }
+        };
+
+        Ok(Self {
+            data_record: None,
+            offset_id,
+            user_id: Some(user_id),
+            final_operation: Operation::Add,
+            metadata_to_be_merged: metadata,
+            final_document: document,
+            final_embedding: embedding,
+        })
+    }
+}
+
+pub(crate) struct LogMaterializer<'a> {
+    // Is None when record segment is uninitialized.
+    record_segment_reader: Option<RecordSegmentReader<'a>>,
+    logs: Chunk<LogRecord>,
+    curr_max_offset_id: Arc<AtomicU32>,
+}
+
+impl<'a> LogMaterializer<'a> {
     pub(crate) fn new(
-        segment_offset_id: u32,
-        log_record: &'a LogRecord,
-        materialized_record: DataRecord<'a>,
+        record_segment_reader: Option<RecordSegmentReader<'a>>,
+        logs: Chunk<LogRecord>,
+        curr_max_offset_id: Arc<AtomicU32>,
     ) -> Self {
         Self {
-            segment_offset_id,
-            log_record,
-            materialized_record,
+            record_segment_reader,
+            logs,
+            curr_max_offset_id,
         }
     }
+    pub(crate) async fn materialize(
+        &'a self,
+    ) -> Result<Chunk<MaterializedLogRecord<'a>>, LogMaterializerError> {
+        // Populate entries that are present in the record segment.
+        let mut existing_id_to_materialized: HashMap<&str, MaterializedLogRecord> = HashMap::new();
+        let mut new_id_to_materialized: HashMap<&str, MaterializedLogRecord> = HashMap::new();
+        // If record segment is uninitialized then there's nothing in the record segment
+        // yet.
+        if self.record_segment_reader.is_some() {
```

Comment:
nit - would prefer we did match switching rather than is_some() unwrap in rust. its always confusing to read longer code that unwraps in response to an if